File: iptv-player/mobile/lib/domain/redact.dart
final secretKey = RegExp(
  r'pass(word|wd)?|secret|token|authorization|cookie|credential|api[-_]?key',
  caseSensitive: false,
);

final _userInfo = RegExp(
  r'([a-z][a-z0-9+.-]*://)([^/\s:@]+):([^/\s@]+)@',
  caseSensitive: false,
);

final _secretQuery = RegExp(
  r'([?&](?:password|passwd|token|secret|apikey|api_key)=)[^&#\s]+',
  caseSensitive: false,
);

const _maxDepth = 6;
const _maxKeys = 40;

String redactText(String value) {
  final withoutUserInfo = value.replaceAllMapped(_userInfo, (match) => '${match[1]}[redacted]@');
  return withoutUserInfo.replaceAllMapped(_secretQuery, (match) => '${match[1]}[redacted]');
}

Object? redact(Object? value, [int depth = 0]) {
  if (depth > _maxDepth) return '[depth]';
  if (value is String) return redactText(value);
  if (value == null || value is num || value is bool) return value;
  if (value is Error || value is Exception) {
    return {'name': value.runtimeType.toString(), 'message': redactText(value.toString())};
  }
  if (value is List) {
    return [for (final item in value) redact(item as Object?, depth + 1)];
  }
  if (value is Map) {
    if (value.length > _maxKeys) return '[object]';
    return {
      for (final entry in value.entries)
        entry.key.toString(): secretKey.hasMatch(entry.key.toString())
            ? '[redacted]'
            : redact(entry.value as Object?, depth + 1),
    };
  }
  return redactText(value.toString());
}

File: iptv-player/shared/redact.ts
const SECRET_KEY = /pass(word|wd)?|secret|token|authorization|cookie|credential|api[-_]?key/i
const USERINFO_PATTERN = /([a-z][a-z0-9+.-]*:\/\/)([^/\s:@]+):([^/\s@]+)@/gi
const SECRET_QUERY_PATTERN = /([?&](?:password|passwd|token|secret|apikey|api_key)=)[^&#\s]+/gi
const MAX_DEPTH = 6
const MAX_KEYS = 40

export function redactText(value: string): string {
  return value
    .replace(USERINFO_PATTERN, '$1[redacted]@')
    .replace(SECRET_QUERY_PATTERN, '$1[redacted]')
}

export function redact(value: unknown, depth = 0): unknown {
  if (depth > MAX_DEPTH) return '[depth]'
  if (typeof value === 'string') return redactText(value)
  if (typeof value !== 'object' || value === null) return value
  if (value instanceof Error) {
    return { name: value.name, message: redactText(value.message) }
  }
  if (Array.isArray(value)) return value.map((item) => redact(item, depth + 1))

  const entries = Object.entries(value)
  if (entries.length > MAX_KEYS) return '[object]'

  const output: Record<string, unknown> = {}
  for (const [key, nested] of entries) {
    output[key] = SECRET_KEY.test(key) ? '[redacted]' : redact(nested, depth + 1)
  }
  return output
}
